core/cap-0063.md

```diff
@@ -0,0 +1,496 @@
+    Authors: Garand Tyson <@SirTyson>
+    Consulted: Dmytro Kozhevin <@dmkozh>, Nicolas Barry <@MonsieurNicolas>
+Status: Draft
+Created: 2024-12-09
+Discussion: https://github.com/stellar/stellar-protocol/discussions/1585
+Protocol version: 23
+```
+
+## Simple Summary
+
+This proposal introduces a new resource type for Soroban reads, distinguishing between in-memory and
+disk reads. This also proposes automatic restoration for archived entries via `InvokeHostFunctionOp`.
+
+## Working Group
+
+As specified in the Preamble.
+
+## Motivation
+
+By distinguishing disk and in-memory reads, this proposal allows for significant increase in Soroban
+read limits. This distinction also allows for safe automatic entry restoration, significantly improving UX.
+
+### Goals Alignment
+
+This change is aligned with the goal of lowering the cost and increasing the scale of the network.
+
+## Abstract
+
+[CAP-0062](cap-0062.md) introduces partial State Archival, where evicted and live Soroban state is stored in separate databases.
+This separation allows live Soroban state to be efficiently cached entirely in memory, removing disc reads entirely for all
+live Soroban state. However, classic state and evicted entries are still subject to disk reads.
+
+This CAP introduces a new resource type for reads, distinguishing between disk reads and in-memory reads. By making this
+distinction at the protocol level, read limits for Soroban data can greatly increase. This also opens the door for other
+optimizations, such as a complete module cache for all live contracts.
+
+Additionally, this CAP introduces automatic restoration via `InvokeHostFunctionOp`, where any archived key present in the
+footprint is automatically restored. Initially in protocol 20, the state archival design was not solidified enough to
+enable automatic restoration, so an explicit restore operation was required. While this was not technically required in
+protocol 20, it was important for contract interfaces to reason properly about restoration such that when full state
+archival was introduce, it would not break preexisting deployments. Given that the full State Archival design has been
+mostly finalized in [CAP-0057](cap-0057.md), it is now appropriate to introduce automatic restoration.
+
+## Specification
+
+### XDR changes
+
+```
+--- a/Stellar-contract-config-setting.x
++++ b/Stellar-contract-config-setting.x
+@@ -26,28 +26,28 @@ struct ConfigSettingContractComputeV0
+ // Ledger access settings for contracts.
+ struct ConfigSettingContractLedgerCostV0
+ {
+-    // Maximum number of ledger entry read operations per ledger
+-    uint32 ledgerMaxReadLedgerEntries;
+-    // Maximum number of bytes that can be read per ledger
+-    uint32 ledgerMaxReadBytes;
++    // Maximum number of live Soroban entry read operations per ledger
++    uint32 ledgerMaxReadLiveSorobanEntries;
++    // Maximum number of live Soroban bytes that can be read per ledger
++    uint32 ledgerMaxLiveSorobanBytes;
```

Comment:
I guess I'm just worried about potential "hidden" costs that could bite us later on if we have no limit. These include things like copies, hashing keys if we do something like a hashset for cache implementation, etc. Given how tightly we've set limits on other resources, it seems like a potential footgun if we allow very large footprints that load many in memory keys. Further down the line, if we do end up needing to limit reads for some unforseen implementation reason, we might break contracts if we apply something retroactively.

That being said, not metering or charging for reads is definitely appealing from a UX perspective, and you're more familiar with resource structure than I am, so I'll defer to you wrt risks of having implicit limits.

Reply:
Also those implicit limits only help us once we're inside the VM. Currently core loads all data prior to entering the VM, metering things like instructions and memory would not apply when loading the footprint. We could potentially Cha get this and vector in and out of the VM for in-memory loads, but that seems like a significant implementation change. In the current load everything and pass it to the VM approach a malicious tx could allocate minimum instructions and memory to reduce the fee but fill the footprint to the max size of 64kb WASM keys to maximize the amount of "free" work.

Reply:
Most of the potential costs are proportional to the count of keys, not the entry size. Given that entries are immutable, I don't see why should we be concerned with the size specifically.

> metering things like instructions and memory would not apply when loading the footprint

'Loading' the footprint is just a bunch of map lookups, isn't it? The metering will kick in as soon as we start decoding the XDR for these (btw it would make sense to store entries as encoded XDR buffers, given that we don't need to interpret them outside of Soroban env). I don't quite see why should this be problematic. And again, I agree that we may still have the limit for the total number of entries accessed per tx and per ledger - the difference is that they're automatically inferred from the footprint, unlike read bytes that need to be accounted for separately.

Reply:
Ah sorry, I originally misread your comment as  "don't limit in-memory reads at all." Ya I agree limiting entry count but not size is fine.